### ISSUE TITLE
rdm only uses 1 cpu even when setting job count

### ISSUE THREAD

**Issue by ismacaulay** [Andersbakken/rtags#529]
What was done:
- start rdm with `rdm -C -j 50` to clean out the project and start it with 50 jobs

What was seen:
- after executing `rc -J <path to compilation db>` rdm is using about 35% of my cpu % (according to htop) and typically only cpu1 is doing any work.

What was expected:
- I would have expected rdm to use all available cpu resources while indexing the compilation database. 

I am using a good computer (intel xeon e5 @ 3.50, 6 cores, 32 gb ram, ssd, Ubuntu 14.04) and the compilation database is ~15000 entries. This causes the indexing to take hours. Let me know if there is anything I do to try to help speed this up.

**Comment by cleeland** [comment]
Are you certain your default process mask permits execution on all cores?

On Friday, November 20, 2015, Ian Macaulay <notifications@github.com> wrote:

> What was done:
>
>    - start rdm with rdm -C -j 50 to clean out the project and start it
>    with 50 jobs
>
> What was seen:
>
>    - after executing rc -J <path to compilation db> rdm is using about
>    35% of my cpu % (according to htop) and typically only cpu1 is doing any
>    work.
>
> What was expected:
>
>    - I would have expected rdm to use all available cpu resources while
>    indexing the compilation database.
>
> I am using a good computer (intel xeon e5 @ 3.50, 6 cores, 32 gb ram, ssd,
> Ubuntu 14.04) and the compilation database is ~15000 entries. This causes
> the indexing to take hours. Let me know if there is anything I do to try to
> help speed this up.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Andersbakken/rtags/issues/529>.
>


-- 
Chris Cleeland

**Comment by ismacaulay** [comment]
I am not 100% sure how to check this, but running `taskset -p <rdm process id>` gives me back a cpu affinity of `fff` which is supposed to mean the the rdm pid can run on any if the 12 cpus my machine has. If there is a different linux command to use let me know.

Other tools such as make, and other build systems that I use, sourceweb (another clang indexer) all use all my cpus, but rdm does not. I have had rdm running for for over an hour and it is only at 12%, and most of my cpus are idle

**Comment by Andersbakken** [comment]
Hi Ian

rdm actually spawns separate processes for the indexing. If you do: $ pidof
rp
You should see the number of jobs running at any given time.

Anders

**Comment by ismacaulay** [comment]
Ahh yes, I see the all the processes for rp, but htop shows most of those processes in a sleeping state. I dont ever really see them jump into a running state. Is that normal? 

The indexing is taking a really long time. Its been over 2 hours and its only 25% done. This just might be a downside of having a project with over 15k targets, but I have used [sourceweb](https://github.com/rprichard/sourceweb) to index the project just to see how clang tools handles large projects. Their indexing tool indexed my project in ~15 min and watching htop it makes use of all 12 of my cpus. I understand that the two indexers are implemented differently and probably index different things, but at the rate its going rtags is going to take almost 8 hours to index my project.

Let me know if you need any more info.

Ian

**Comment by cslux** [comment]
@ismacaulay Yeah, there must be something wrong.

I build a target (content_shell from chromium) containting 13k translation units, which takes a bit more than 1 hour on my laptop.
- Intel(R) Core(TM) i7-4710MQ CPU @ 2.50GHz
- 24 GB RAM
- PCI SSD

You project should be build with rtags within 15-25 minutes on your hardware.

Christian

**Comment by ismacaulay** [comment]
@cslux Its good to know there is something to reference it too. I do agree something is wrong.

I have not had time to dig into the code, but watching htop, it almost feels like something is blocking the the processes such as a clang error. As I watched htop, the processes would sit sleeping, then when I think the indexing was complete, they would hammer all 12 cores for a second then go back to idle. While indexing it does print some errors as it goes about things it can't find, or flags it doesn't understand.

One thing that could be an issue is that the compiler that is in the compilation database commands is not the actual compiler it was compiled with. We compile with an old version of gcc via a schroot jail (so rtags does not actually have access to it), which means rtags will actually be using the system gcc (4.8 i think). I am not sure if that is a problem or not as I do not know enough about the libclang libraries, but I would not have expected it to block the processes, unless there is some error handling code within rtags that could be causing this. If I have some time on next week, I will dig into the source and see what I can find (any suggestions on a starting place are welcome)

Ian

**Comment by Andersbakken** [comment]
Hi Ian

Can you build rtags in debug and try to attach to one of the rp processes?
Stricktly speaking one would expect it to be spend the majority of time
inside clang_parseTranslationUnit2 and most of the rest of it inside
clang_visitChildren.

If you find it sitting inside EventLoop::exec (would be happening from
underneath clang_visitChildren) a lot that would be indicative of rp not
being able to communicate back to rdm properly (or at least swiftly).

A couple of spot checks could be a good idea to get a handle on what went
wrong.

Also can you send some of the output that rdm prints while things are
indexing? There's some timing information in there.

E.g.

[ 29%] 2/7 18:41:58 ~/dev/rtags/src/rp.cpp in 3642ms. (548 syms, 647
symNames, 21 includes, 2 of 607 files, symbols: 686 of 32674, 1 queried)
(3598/36/4ms) (dirty). (priority 5)
[ 43%] 3/7 18:41:59 ~/dev/rtags/src/JobScheduler.cpp in 3902ms. (1126 syms,
265 symNames, 3 includes, 1 of 605 files, symbols: 1737 of 33673, 0
queried) (3851/43/5ms) (dirty). (priority 5)
[ 57%] 4/7 18:41:59 ~/dev/rtags/src/ClangIndexer.cpp in 3924ms. (3615 syms,
1590 symNames, 13 includes, 1 of 606 files, symbols: 5024 of 36904, 1
queried) (3842/66/13ms) (dirty). (priority 5)
[ 71%] 5/7 18:42:00 ~/dev/rtags/src/rct/rct/SocketServer.cpp in 1421ms.
(313 syms, 148 symNames, 14 includes, 1 of 251 files, symbols: 435 of
16643, 0 queried) (1408/10/2ms) (dirty). (priority 5)
[ 86%] 6/7 18:42:00 ~/dev/rtags/src/Project.cpp in 5675ms. (5651 syms, 2310
symNames, 21 includes, 1 of 614 files, symbols: 7790 of 39845, 0 queried)
(5574/83/15ms) (dirty). (priority 5)
[100%] 7/7 18:42:01 ~/dev/rtags/src/rct/rct/Connection.cpp in 2021ms. (564
syms, 186 symNames, 8 includes, 1 of 247 files, symbols: 797 of 16415, 0
queried) (2001/15/4ms) (dirty). (priority 3)

The numbers at the end (%d/%d/%dms) indicate the time spent in
clang_parseTranslationUnit and in clang_visitChildren. The last number is
rp communicating back to rdm.

Anders

**Comment by Andersbakken** [comment]
Attach with gdb I meant.

Anders

**Comment by ismacaulay** [comment]
Hi Anders,

Here is some of the output from rtags, I had to sanitize the output since the project is proprietary. When I have some more time, I will do some debugging with gdb.

```
[ 97%] 14866/15349 17:01:02 <file being processed> in 55071ms. (94 syms, 74 symNames, 2 includes, 2 of 65 files, symbols: 104 of 3551, 1 queried) (220/54832/1ms). (priority 0)
[ 97%] 14867/15349 17:01:03 <file being processed> in 55088ms. (123 syms, 200 symNames, 9 includes, 2 of 176 files, symbols: 264 of 11644, 1 queried) (806/54263/1ms). (priority 0)
[ 97%] 14868/15349 17:01:05 <file being processed> in 39969ms. (200 syms, 169 symNames, 9 includes, 2 of 82 files, symbols: 241 of 3398, 1 queried) (186/39767/1ms). (priority 0)
[ 97%] 14869/15349 17:01:07 <file being processed> in 40010ms. (638 syms, 265 symNames, 23 includes, 2 of 188 files, symbols: 851 of 7458, 1 queried) (437/39549/2ms). (priority 0)
[ 97%] 14870/15349 17:01:09 <file being processed> in 174ms. (42 syms, 28 symNames, 7 includes, 1 of 43 files, symbols: 52 of 2099, 0 queried) (155/2/0ms). (priority 0)
[ 97%] 14871/15349 17:01:11 <file being processed> in 136ms. (90 syms, 98 symNames, 5 includes, 1 of 37 files, symbols: 96 of 2027, 0 queried) (112/4/2ms). (priority 0)
[ 97%] 14872/15349 17:01:13 <file being processed> in 125ms. (25 syms, 26 symNames, 4 includes, 1 of 43 files, symbols: 26 of 2052, 0 queried) (110/2/0ms). (priority 0)
[ 97%] 14873/15349 17:01:15 <file being processed> in 169ms. (211 syms, 146 symNames, 1 includes, 1 of 58 files, symbols: 265 of 3493, 0 queried) (149/4/1ms). (priority 0)
[ 97%] 14874/15349 17:01:17 <file being processed> in 301ms. (4 syms, 6 symNames, 1 includes, 1 of 62 files, symbols: 3 of 7979, 0 queried) (278/6/0ms). (priority 0)
[ 97%] 14875/15349 17:01:19 <file being processed> in 35ms. (9 syms, 12 symNames, 1 includes, 1 of 9 files, symbols: 9 of 554, 0 queried) (20/1/1ms). (priority 0)
[ 97%] 14877/15350 17:01:20 <file being processed> in 45666ms. (689 syms, 640 symNames, 17 includes, 2 of 488 files, symbols: 1206 of 16725, 1 queried) (1625/44023/2ms). (priority 0)
[ 97%] 14878/15350 17:01:22 <file being processed> in 45395ms. (1683 syms, 1403 symNames, 36 includes, 2 of 830 files, symbols: 2854 of 22522, 1 queried) (1954/43418/4ms). (priority 0)
[ 97%] 14879/15350 17:01:24 <file being processed> in 45338ms. (679 syms, 744 symNames, 24 includes, 2 of 828 files, symbols: 1090 of 20923, 1 queried) (2191/43126/2ms). (priority 0)
[ 97%] 14880/15350 17:01:26 <file being processed> in 67714ms. (159 syms, 221 symNames, 5 includes, 3 of 105 files, symbols: 309 of 5351, 2 queried) (342/67353/2ms). (priority 0)
[ 97%] 14881/15350 17:01:28 <file being processed> in 69608ms. (346 syms, 299 symNames, 15 includes, 3 of 497 files, symbols: 538 of 10640, 2 queried) (871/68721/2ms). (priority 0)
Output from <file being processed>:
<path>:50:10: fatal error: 'iterator' file not found
Error getting fileId for <path> inclusion directive <class> <path>:14:1 60060 60000

Output from <file being processed>:
/usr/include/features.h:374:12: fatal error: 'sys/cdefs.h' file not found
Error getting fileId for <path> macro expansion <macro> <path>:157:9 60060 60000

Output from <file being processed>:
<path>:9:10: fatal error: 'functional' file not found
couldn't create included file inclusion directive i functional <path>:9:1
Error getting fileId for <path> macro expansion <macro> <path>:675:7 60059 60000

Output from <file being processed>:
/usr/include/features.h:374:12: fatal error: 'sys/cdefs.h' file not found

Output from <file being processed>:
/usr/include/features.h:374:12: fatal error: 'sys/cdefs.h' file not found

Output from <file being processed>:
/usr/include/features.h:374:12: fatal error: 'sys/cdefs.h' file not found

Output from <file being processed>:
/usr/include/features.h:374:12: fatal error: 'sys/cdefs.h' file not found

Output from <file being processed>:
/usr/include/features.h:374:12: fatal error: 'sys/cdefs.h' file not found

Output from <file being processed>:
/usr/include/features.h:374:12: fatal error: 'sys/cdefs.h' file not found

Output from <file being processed>:
<path>:9:10: fatal error: 'initializer_list' file not found

Output from <file being processed>:
/usr/include/features.h:374:12: fatal error: 'sys/cdefs.h' file not found

Output from <file being processed>:
<path>:96:10: fatal error: 'algorithm' file not found

Output from <file being processed>:
/usr/include/features.h:374:12: fatal error: 'sys/cdefs.h' file not found

Output from <file being processed>:
/usr/include/features.h:374:12: fatal error: 'sys/cdefs.h' file not found

Output from <file being processed>:
/usr/include/features.h:374:12: fatal error: 'sys/cdefs.h' file not found

Output from <file being processed>:
/usr/include/features.h:374:12: fatal error: 'sys/cdefs.h' file not found

Output from <file being processed>:
/usr/include/features.h:374:12: fatal error: 'sys/cdefs.h' file not found

Output from <file being processed>:
/usr/include/features.h:374:12: fatal error: 'sys/cdefs.h' file not found

[ 97%] 14882/15350 17:01:31 <file being processed> in 179ms. (14 syms, 18 symNames, 1 includes, 1 of 58 files, symbols: 13 of 3219, 0 queried) (154/4/1ms). (priority 0)
[ 97%] 14883/15350 17:01:33 <file being processed> in 123ms. (58 syms, 58 symNames, 5 includes, 1 of 52 files, symbols: 72 of 2424, 0 queried) (103/6/1ms). (priority 0)
[ 97%] 14884/15350 17:01:35 <file being processed> in 98ms. (39 syms, 42 symNames, 5 includes, 1 of 31 files, symbols: 44 of 2046, 0 queried) (76/2/1ms). (priority 0)
[ 97%] 14885/15350 17:01:37 <file being processed> in 217ms. (63 syms, 82 symNames, 6 includes, 1 of 110 files, symbols: 67 of 3771, 0 queried) (187/4/0ms). (priority 0)
[ 97%] 14886/15350 17:01:39 <file being processed> in 63611ms. (867 syms, 777 symNames, 18 includes, 3 of 203 files, symbols: 1699 of 8840, 2 queried) (475/63112/2ms). (priority 0)
[ 97%] 14887/15350 17:01:41 <file being processed> in 66818ms. (401 syms, 340 symNames, 9 includes, 3 of 111 files, symbols: 719 of 5782, 2 queried) (322/66475/3ms). (priority 0)
[ 97%] 14890/15352 17:01:43 <file being processed> in 34437ms. (6 syms, 10 symNames, 2 includes, 2 of 51 files, symbols: 4 of 6631, 1 queried) (265/34155/1ms). (priority 0)
[ 97%] 14891/15352 17:01:44 <file being processed> in 34644ms. (107 syms, 190 symNames, 9 includes, 2 of 171 files, symbols: 238 of 11211, 1 queried) (516/34109/2ms). (priority 0)
[ 97%] 14893/15353 17:01:46 <file being processed> in 37673ms. (99 syms, 197 symNames, 9 includes, 2 of 269 files, symbols: 229 of 13543, 1 queried) (675/36978/2ms). (priority 0)
[ 97%] 14895/15354 17:01:48 <file being processed> in 78460ms. (1254 syms, 611 symNames, 8 includes, 3 of 116 files, symbols: 1908 of 7012, 2 queried) (437/77999/2ms). (priority 0)
[ 97%] 14896/15354 17:01:50 <file being processed> in 129263ms. (2476 syms, 1367 symNames, 59 includes, 5 of 655 files, symbols: 3210 of 16987, 4 queried) (1134/128110/4ms). (priority 0)
```

**Comment by Andersbakken** [comment]
Sorry about the delay. There's an unusual amount of time spent in visiting the AST after parsing it. E.g. 516/34109/2ms means that clang used only 516 ms to parse the file whereas rp used 34 seconds to visit the cursors.

I believe we have a problem communicating back to rdm over the unix socket and that it times out whenever it tries to query back. 

Are rc commands working? 

E.g. Can you run:

rc -w 

and get the output in a reasonable amount of time? 

I also added some additional timing to the output but it very much seems like the delay is a direct function of the number of files queried.

In short rp has to "phone home" to rdm whenever we see a file that hasn't been indexed yet.

**Comment by ismacaulay** [comment]
No worries about the delay. It gave me some time to do a little experimenting myself. So I tried using gdb on one of the rp processes, and was less than successful on getting anywhere (I was trying to learn gdb while doing this so I was probably making mistakes along the way). Anyway, I grabed another one of our projects that uses the same environment, but is known to build cleanly with clang. It works just fine. All cores are used and the project indexes quickly (a much smaller project). The larger project does not build cleanly with clang (I am currently working on that) so I am not sure if there is just some code that is causing the clang api itself to be slow. I will keep investigating when I have some time (which is usually only a few hours a week), and keep playing with gdb to help narrow down the problem.

For both projects rc commands work well. I was using a modified version of the sublime-rtags plugin, and got completion results, and references in a timely fashion.

I will pull master, and see if the additional timing data provides any insight later this week as well.

**Comment by Andersbakken** [comment]
Sounds good. Let me know what you find.

Anders

**Comment by dawiddr** [comment]
I've had similar problems when indexing a large project: it would take several hours with low CPU usage, `rp` processes basically waiting idle. But what I've also noticed was that `rdm` was doing lots of IO. So I've moved my `.rtags` directory to a ram disk and that has improved the situation significantly - indexing is now faster by an order of magnitude.

Of course a ram disk comes with its own disadvantages, so for me that's only a workaround. I'm wondering though why `rdm` has to write that much data. My `.rtags` directory is 3 GB is size, but during indexing `rdm` process has written about 600 GB in total.

**Comment by Andersbakken** [comment]
Sorry about the long delay. 

RTags is rather heavy on IO but those numbers sounds like way too much. How do you measure it? 

I just added something to keep track of how many bytes are written per project. You can access the data using:

rc -s project

**Comment by dawiddr** [comment]
Thanks for the counter. I've reindexed the project from scratch and this is what I'm getting:
```
rc -s project
Bytes written: 6364878832
```
6.4 GB looks much more reasonable when compared with `.rtags` directory size, which is 4.6 GB.
600 GB comes from OS X Activity Monitor and unfortunately I can't find what exactly it includes.

**Comment by Andersbakken** [comment]
Hm. Maybe it possibly counts our mmap'ing as writes?

We do open the file descriptor with O_RDONLY and mmap with PROT_READ but
still.

Are you running with the log file enabled by any chance? We log kind of
obscene amounts.

Anders
